File: PoliTechAPI/pt-api/src/main/java/ru/pt/api/service/calculator/CalculatorService.java
package ru.pt.api.service.calculator;

import ru.pt.api.dto.calculator.CalculatorModel;
import ru.pt.api.dto.calculator.CalculatorTemplateLine;
import ru.pt.api.dto.calculator.CalculatorTemplate;
import ru.pt.domain.model.CalculatorContext;
import java.util.List;

/**
 * Методы для CRUD операций с калькуляторами и для расчета страховой премии договора
 */
public interface CalculatorService {

    /**
     * Получить сохраненный калькулятор продукта
     * @param productId айди продукта
     * @param versionNo номер версии
     * @param packageNo номер пакета
     * @return модель калькулятора или null, если не найден
     */
    CalculatorModel getCalculator(Long tenantId, Long productId, Long versionNo, String packageNo);

    /**
     * Создать калькулятор, если он отсутствует
     * @param productId айди продукта
     * @param productCode код продукта
     * @param versionNo номер версии
     * @param packageNo номер пакета
     * @return созданный или найденный калькулятор
     */
    CalculatorModel createCalculator(Long tenantId, Long productId, Long versionNo, String packageNo, Long templateId);

    /**
     * копирует калькулятор в другой продукт версию
     * @param tenantId айди тенанта
     * @param productId айди продукта
     * @param versionNo номер версии
     * @param packageNo номер пакета
     * @return модель калькулятора
     */
    public void copyCalculator(Long tenantId, Long productId, Long versionNo, String packageNo, Long versionNoTo) ;

    /**
     * Выполнить калькулятор и вернуть рассчитанные переменные
     * @param tenantId айди тенанта
     * @param productId айди продукта
     * @param versionNo номер версии
     * @param packageNo номер пакета
     * @param inputValues входные переменные
     * @return список переменных с рассчитанными значениями
     */
    void runCalculator(Long tenantId, Long productId, Long versionNo, String packageNo, CalculatorContext ctx);

    /**
     * Заменить модель калькулятора
     * @param tenantId айди тенанта
     * @param productId айди продукта
     * @param productCode код продукта
     * @param versionNo номер версии
     * @param packageNo номер пакета
     * @param newJson новая модель калькулятора
     * @return сохраненная модель
     */
    CalculatorModel replaceCalculator(Long tenantId, Long productId, String productCode, Long versionNo,
                                      String packageNo, CalculatorModel newJson);

    /**
     * Синхронизировать переменные калькулятора с продуктом
     * @param calculatorId идентификатор калькулятора
     */
    void syncVars(Long tenantId, Long calculatorId);

    /**
     * Получить калькулятор по идентификатору
     * @param calculatorId идентификатор калькулятора
     * @return модель калькулятора или null, если не найден
     */
    CalculatorModel getCalculatorById(Long tenantId, Long calculatorId);

    void deleteCalculator(Long tenantId, Long productId, Long versionNo, String packageNo);

    /**
     * Создать шаблон формулы по существующему калькулятору для заданной линии бизнеса.
     */
    CalculatorTemplate createTemplate(Long tenantId, String lobCode, Long calculatorId);

    /**
     * Получить все шаблонные строки формул для линии бизнеса.
     */
    List<CalculatorTemplate> getTemplates(Long tenantId, String lobCode);

    /**
     * Создать шаблон формулы по существующему калькулятору для заданной линии бизнеса.
     */
    CalculatorTemplate updateTemplateName(Long tenantId, Long templateId, String templateName);

    /**
     * Удалить шаблон формулы
     * @param tenantId айди тенанта
     * @param templateId идентификатор шаблона
     */
    void deleteTemplate(Long tenantId, Long templateId);

}

File: PoliTechFront/src/app/shared/services/calculator.service.ts
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { catchError, delay, tap } from 'rxjs/operators';
import { AuthService } from './auth.service';
import { BaseApiService } from './api/base-api.service';
import { EnvService } from './env.service';

export interface Calculator {
  id?: string;
  productId: string;
  productCode: string;
  versionNo: string;
  packageNo: string;
  vars: CalculatorVar[];
  formulas: CalculatorFormula[];
  coefficients: CalculatorCoefficient[];
}

export interface CalculatorVar {
  varCode: string;
  varName: string;
  varPath: string;
  varType: string;
  varValue: string;
  varDataType: string;
  varCdm: string;
  varNr: number;
}



export interface CalculatorFormula {
  varCode: string;
  varName: string;
  lines: FormulaLine[];
}

export interface FormulaLine {
  nr: string;
  conditionLeft: string;
  conditionOperator: string;
  conditionRight: string;
  expressionResult: string;
  expressionLeft: string;
  expressionOperator: string;
  expressionRight: string;
  postProcessor?: string;
}

export interface VarReference1 {
  varCode: string;
  varName: string;
}

export interface CalculatorCoefficient {
  varCode: string;
  varName: string;
  altVarName?: string;
  altVarValue?: number;
  columns: CoefficientColumn[];
  errorTextIfNotFound?: string;
}

export interface CoefficientColumn {
  varCode: string;
  varDataType: string;
  nr: string;
  conditionOperator: string;
  sortOrder: string;
}

export interface CoefficientDataRow {
  id: number;
  conditionValue: (string | null)[];
  resultValue: number;
}

/** Ответ POST `/admin/calculators/templates` — строки шаблона формулы */
export interface CalculatorTemplateLine {
  calculatorId?: number;
  lineNr?: number;
  text?: string;
}

export interface CalculatorTemplate {
  calculatorId?: number;
  calculatorName?: string;
  lines: CalculatorTemplateLine[];
}

@Injectable({
  providedIn: 'root'
})

export class CalculatorService extends BaseApiService<Calculator> {
  constructor(
    http: HttpClient,
    env: EnvService,
    authService: AuthService,
  ) {
    super(http, env, 'admin/calculators', authService);
  }

  private mockData: Calculator = {
    productId: '',
    productCode: '',
    versionNo: '',
    packageNo: '',
    vars: [],
    formulas: [],
    coefficients: []
  };

  // GET calculator
  getCalculator(productId: string, versionNo: string, packageNo: string): Observable<Calculator> {

    let baseUrl = this.getUrl() + '/products/' + productId + '/versions/' + versionNo + '/packages/' + packageNo;
    const headers = new HttpHeaders({ 'X-Skip-Global-Error': 'true' });
    return this.http.get<Calculator>(baseUrl, { headers }).pipe(
      tap(data => {
        this.mockData = data;
      }),
      catchError(error => {
        console.error('Error fetching calculator:', error);
        throw error;
      })
    );
  }

  // POST calculator (create new), optional from template
  createCalculator(productId: string, versionNo: string, packageNo: string, templateId?: number): Observable<Calculator> {
    const body = {
      productId: productId,
      versionNo: versionNo,
      packageNo: packageNo,
      templateId: templateId ?? null
    };

    let baseUrl = this.getUrl() + '/products/' + productId + '/versions/' + versionNo + '/packages/' + packageNo;
      return this.http.post<Calculator>(baseUrl, body).pipe(
        tap(createdCalculator => {
          this.mockData = { ...createdCalculator };
        }),
        catchError(error => {
          console.error('Error creating calculator:', error);
          this.mockData = { ...this.mockData, ...body };
          return of({ ...this.mockData });
        })
      );
    

    this.mockData = { ...this.mockData, ...body };
    return of({ ...this.mockData }).pipe(delay(500));
  }

  // PUT calculator (update)
  updateCalculator(calculator: Calculator): Observable<Calculator> {
    
    let baseUrl = this.getUrl() + '/products/' + calculator.productId + '/versions/' + calculator.versionNo + '/packages/' + calculator.packageNo;
      return this.http.put<Calculator>(baseUrl, calculator).pipe(
        tap(updatedCalculator => {
          this.mockData = { ...updatedCalculator };
        }),
        catchError(error => {
          console.error('Error updating calculator:', error);
          return of({ ...this.mockData });
        })
      );
    

    this.mockData = { ...calculator };
    return of({ ...this.mockData }).pipe(delay(500));
  }

  // Mock data for dropdowns
  getVarTypeOptions(): Observable<string[]> {
    return of(['IN', 'OUT', 'CALC']).pipe(delay(200));
  }

  getConditionOperatorOptions(): Observable<string[]> {
    return of(['=', '!=', '>', '<', '>=', '<=', 'IN', 'NOT_IN']).pipe(delay(200));
  }

  getExpressionOperatorOptions(): Observable<string[]> {
    return of(['+', '-', '*', '/', 'min', 'max']).pipe(delay(200));
  }

  getPostProcessorOptions(): Observable<string[]> {
    return of(['round2_up', 'round2_down', 'round2_ceiling', 'round2_floor'
      , 'round2_halfup', 'round2_halfdown', 'round2_halfeven'
      , 'NONE']).pipe(delay(200));

  }

  getSortOrderOptions(): Observable<string[]> {
    return of(['ASC', 'DESC','NONE']).pipe(delay(200));
  }

  // Coefficient data API
  getCoefficientData(calculatorId: string, coefficientCode: string): Observable<CoefficientDataRow[]> {
    let baseUrl = this.getUrl(calculatorId) + '/coefficients/' + coefficientCode;

    return this.http.get<CoefficientDataRow[]>(baseUrl).pipe(
      catchError(error => {
        console.error('Error fetching coefficient data:', error);
        // Return mock empty data
        return of([]);
      })
    );
  }

  createCoefficientData(calculatorId: string, coefficientCode: string, rows: CoefficientDataRow[]): Observable<CoefficientDataRow[]> {
    let baseUrl = this.getUrl(calculatorId) + '/coefficients/' + coefficientCode;

    return this.http.post<CoefficientDataRow[]>(baseUrl, rows).pipe(
      catchError(error => {
        console.error('Error creating coefficient data:', error);
        return of(rows);
      })
    );
  }

  updateCoefficientData(calculatorId: string, coefficientCode: string, rows: CoefficientDataRow[]): Observable<CoefficientDataRow[]> {
    let baseUrl = this.getUrl(calculatorId) + '/coefficients/' + coefficientCode;

    return this.http.put<CoefficientDataRow[]>(baseUrl, rows).pipe(
      catchError(error => {
        console.error('Error updating coefficient data:', error);
        return of(rows);
      })
    );
  }

  /** GET coefficient SQL template (variable names as placeholders) */
  getCoefficientSql(calculatorId: string, coefficientCode: string): Observable<string> {
    const url = this.getUrl(calculatorId) + '/coefficients/' + encodeURIComponent(coefficientCode) + '/SQL';
    return this.http.get(url, { responseType: 'text' }).pipe(
      catchError(error => {
        console.error('Error fetching coefficient SQL:', error);
        throw error;
      })
    );
  }

  /** POST: создать шаблон формулы по калькулятору для LOB (бэкенд: createTemplate) */
  createTemplate(lobCode: string, calculatorId: string | number): Observable<CalculatorTemplate> {
    const body = {
      lobCode,
      calculatorId: typeof calculatorId === 'string' ? Number(calculatorId) : calculatorId,
    };
    return this.http.post<CalculatorTemplate>(`${this.getUrl()}/templates`, body).pipe(
      catchError(error => {
        console.error('Error creating calculator template:', error);
        throw error;
      })
    );
  }

  /** GET: получить шаблоны калькуляторов для LOB */
  getTemplates(lobCode: string): Observable<CalculatorTemplate[]> {
    const params = new HttpParams().set('lob', lobCode);
    return this.http.get<CalculatorTemplate[]>(`${this.getUrl()}/templates`, { params }).pipe(
      catchError(error => {
        console.error('Error fetching calculator templates:', error);
        throw error;
      })
    );
  }

  /** PUT: обновить имя шаблона калькулятора */
  updateTemplateName(templateId: number, templateName: string): Observable<CalculatorTemplate> {
    return this.http.put<CalculatorTemplate>(`${this.getUrl()}/templates/${templateId}`, { templateName }).pipe(
      catchError(error => {
        console.error('Error updating calculator template name:', error);
        throw error;
      })
    );
  }

  /** DELETE: удалить шаблон калькулятора */
  deleteTemplate(templateId: number): Observable<void> {
    return this.http.delete<void>(`${this.getUrl()}/templates/${templateId}`).pipe(
      catchError(error => {
        console.error('Error deleting calculator template:', error);
        throw error;
      })
    );
  }
}
